skfda/preprocessing/dim_reduction/projection/_fpca.py

```diff
@@ -1,33 +1,445 @@
+            X (FDataBasis):
+                the functional data object to be analysed
+            y (None, not used):
+                only present because of fit function convention
+
+        Returns:
+            (array_like): the scores of the data with reference to the
+            principal components
+        """
+
+        # in this case it is the inner product of our data with the components
+        return X.inner_product(self.components_)
+
+
+class FPCAGrid(FPCA):
+    """Funcional principal component analysis for functional data represented
+    in discretized form.
+
+    Attributes:
+        components_ (FDataBasis): this contains the principal components either
+            in a basis form.
+        component_values_ (array_like): this contains the values (eigenvalues)
+            associated with the principal components.
+        pca_ (sklearn.decomposition.PCA): object for principal component analysis.
+            In both cases (discretized FPCA and basis FPCA) the problem can be
+            reduced to a regular PCA problem and use the framework provided by
+            sklearn to continue.
+
+    Parameters:
+        n_components (int): number of principal components to obtain from
+            functional principal component analysis. Defaults to 3.
+        centering (bool): if True then calculate the mean of the functional data
+            object and center the data first. Defaults to True. If True the
+            passed FDataBasis object is modified.
+        weights (numpy.array): the weights vector used for discrete
+            integration. If none then the trapezoidal rule is used for
+            computing the weights.
+
+    Examples:
+        In this example we apply discretized functional PCA with some simple
+        data to illustrate the usage of this class. We initialize the
+        FPCADiscretized object, fit the artificial data and obtain the scores.
+        The results are not tested because there are several equivalent
+        possibilities.
+
+        >>> data_matrix = np.array([[1.0, 0.0], [0.0, 2.0]])
+        >>> sample_points = [0, 1]
+        >>> fd = FDataGrid(data_matrix, sample_points)
+        >>> fpca_grid = FPCAGrid(2)
+        >>> fpca_grid = fpca_grid.fit(fd)
     """
-    fdatagrid = fdatagrid - mean(fdatagrid)  # centers the data
-    # singular value decomposition
-    u, s, v = np.linalg.svd(fdatagrid.data_matrix)
-    principal_directions = v.T  # obtain the eigenvectors matrix
-    eigenvalues = (np.diag(s) ** 2) / (fdatagrid.n_samples - 1)
-    scores = u @ s  # functional principal scores
-
-    return scores, principal_directions, eigenvalues
+
+    def __init__(self, n_components=3, weights=None, centering=True):
+        """FPCABasis constructor
+
+        Args:
+            n_components (int): number of principal components to obtain from
+                functional principal component analysis
+            weights (numpy.array): the weights vector used for discrete
```

Comment:
I think this could also be a callable.

Reply:
Calling FPCAGrid on a FDataGrid object? with default parameters or the user may also choose the parameters?